config/configcheck/configcheck.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package configcheck has checks to be applied to configuration
+// objects implemented by factories of components used in the OpenTelemetry
+// collector. It is recommended for implementers of components to run the
+// validations available on this package.
+package configcheck
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"strings"
+
+	"github.com/open-telemetry/opentelemetry-collector/config"
+	"github.com/open-telemetry/opentelemetry-collector/oterr"
+)
+
+// The regular expression for valid config field tag.
+var configFieldTagRegExp = regexp.MustCompile("^[a-z0-9][a-z0-9_]*$")
+
+// ValidateConfigFromFactories checks if all configurations for the given factories
+// are satisfying the patterns used by the collector.
+func ValidateConfigFromFactories(factories config.Factories) error {
+	var errs []error
+	var configs []interface{}
+
+	for _, factory := range factories.Receivers {
+		configs = append(configs, factory.CreateDefaultConfig())
```

Comment:
Since CreateDefaultConfig is invoked in Load Config (), to save on duplicating that call - why not invoke `ValidateConfig` in load(extensions/processors/...) there.

Reply:
Yeah, crossed my mind but I opted to have it this way for a few two reasons:

1. I wanted it to be an error on its own not mixed with the actual start;
2. Avoiding spreading the calls over the code;
3. Most important: I wanted the code to check all factories not only the ones in the current configuration.

Reply:
I agree that 3 is most important benefit of this approach.

Reply:
Why don't we make it a proper test case that we check all of these. That way we catch it during impl/testing and then we can move to invoking the call in the Load.

And we can make a future task to separate config validation/load config and service initialization in a future task. That would also make the validation flag even neater.

Reply:
There is a test case loading all the config but we need also to have it running on the collector otherwise components not following the pattern can escape the check: someone building a collector with a custom set of components would not go through the same validation.

The idea is also to recommend implementers of components to use it directly in their tests, that's why the `ValidationConfig` function:  they can catch issues on their own before having it running under the collector. If this check is 100% separated from the normal run it is possible something that runs but doesn't pass the validation, it is better to fail to launch in this case.

Reply:
I will add the advice to use to the create function on the interfaces, and to all of our own components so anyone using them as base will see the usage of `ValidateConfig`

Reply:
Done.